service/frontend/versionChecker.go

```diff
@@ -0,0 +1,155 @@
+package frontend
+
+import (
+	"go.temporal.io/api/serviceerror"
+	"go.temporal.io/server/api/persistenceblobs/v1"
+	"go.temporal.io/server/common/metrics"
+	"go.temporal.io/server/common/persistence"
+	"go.temporal.io/server/common/primitives/timestamp"
+	"runtime"
+	"sync"
+	"time"
+
+	"go.temporal.io/server/common/headers"
+	"go.temporal.io/server/common/resource"
+	"go.temporal.io/version/check"
+)
+
+const VersionCheckInterval = 24 * time.Hour
+
+type VersionChecker struct {
+	resource.Resource
+	params       *resource.BootstrapParams
+	config       *Config
+	shutdownChan chan struct{}
+	metricsScope metrics.Scope
+	startOnce    sync.Once
+	stopOnce     sync.Once
+}
+
+func NewVersionChecker(
+	resource resource.Resource,
+	params *resource.BootstrapParams,
+	config *Config,
+) *VersionChecker {
+	return &VersionChecker{Resource: resource, params: params, config: config, shutdownChan: make(chan struct{}),
+		metricsScope: resource.GetMetricsClient().Scope(metrics.VersionCheckScope)}
+}
+
+func (vc *VersionChecker) Start() {
+	if vc.config.EnableServerVersionCheck() {
+		vc.startOnce.Do(func() {
+			go vc.versionCheckLoop()
+		})
+	}
+}
+
+func (vc *VersionChecker) Stop() {
+	if vc.config.EnableServerVersionCheck() {
+		vc.stopOnce.Do(func() {
+			close(vc.shutdownChan)
+		})
+	}
+}
+
+func (vc *VersionChecker) versionCheckLoop() {
+	timer := time.NewTicker(VersionCheckInterval)
+	defer timer.Stop()
+	vc.performVersionCheck()
+	for {
+		select {
+		case <-vc.shutdownChan:
+			return
+		case <-timer.C:
+			vc.performVersionCheck()
+		}
+	}
+}
+
+func (vc *VersionChecker) performVersionCheck() {
+	sw := vc.metricsScope.StartTimer(metrics.VersionCheckLatency)
+	defer sw.Stop()
+	req, err := vc.createVersionCheckRequest()
+	if err != nil {
+		vc.metricsScope.IncCounter(metrics.VersionCheckFailedCount)
+		return
+	}
+	resp, err := vc.getVersionInfo(req)
+	if err != nil {
+		vc.metricsScope.IncCounter(metrics.VersionCheckFailedCount)
+		return
+	}
+	err = vc.saveVersionInfo(resp)
+	if err != nil {
+		vc.metricsScope.IncCounter(metrics.VersionCheckFailedCount)
+		return
+	}
+	vc.metricsScope.IncCounter(metrics.VersionCheckSuccessCount)
+}
+
+func (vc *VersionChecker) createVersionCheckRequest() (*check.VersionCheckRequest, error) {
+	clusterMetadataManager := vc.GetClusterMetadataManager()
+	metadata, err := clusterMetadataManager.GetClusterMetadata()
+	if err != nil {
+		return nil, err
+	}
+	return &check.VersionCheckRequest{
+		Product:   headers.ClientNameServer,
+		Version:   headers.ServerVersion,
+		Arch:      runtime.GOARCH,
+		OS:        runtime.GOOS,
+		DB:        clusterMetadataManager.GetName(),
+		ClusterID: metadata.ClusterId,
+		Timestamp: time.Now().UnixNano(),
+	}, nil
+}
+
+func (vc *VersionChecker) getVersionInfo(req *check.VersionCheckRequest) (*check.VersionCheckResponse, error) {
+	return check.NewCaller().Call(req)
```

Comment:
So, this will never panic?

Reply:
It shouldn't, if it doesn't it's fine too as we run in sync.Once